File: java/290 Word Pattern/WordPattern.java

/**
 * Given a pattern and a string str, find if str follows the same pattern.
 * Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty word in str.
 */
public class Solution {
	public boolean wordPattern(String pattern, String str) {
		String[] data = str.split(" ");
		if (pattern.length() != data.length)
			return false;
        java.util.Map<Character, String> map = new java.util.HashMap<Character, String>();
		for (int i = 0; i < pattern.length(); i++) {
			char key = pattern.charAt(i);
			String s = map.get(key);
			if (s == null) {
				if (map.containsValue(data[i]))
					return false;
				map.put(key, data[i]);
			} else if (!s.equals(data[i])) {
				return false;
			}
		}
		return true;
	}
}

File: py/290 Word Pattern/WordPattern.py

# Given a pattern and a string str, find if str follows the same pattern.
# Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty word in str.
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        data = str.split(' ')
        if len(pattern) != len(data):
            return False
        map = {}
        for key, data_i in zip(pattern, data):
            s = map.get(key)
            if s == None:
                if data_i in map.values():
                    return False
                map[key] = data_i
            elif not s == data_i:
                return False
        return True
